frontend/benefit/applicant/src/components/applications/forms/application/utils.ts

```diff
@@ -1,8 +1,42 @@
 import { Application } from 'benefit-shared/types/application';
 import { FormikErrors, FormikTouched } from 'formik';
 
-export const getTouchedAndInvalidFields = (
+type FlattenedObject = Record<string, string>;
+
+function flattenObject(
+  obj: FlattenedObject,
+  parentKey = '',
+  result: FlattenedObject = {}
+): FlattenedObject {
+  const appendResult = result;
+  Object.keys(obj).forEach((key) => {
+    const newKey = `${parentKey}${parentKey ? '.' : ''}${key}`;
+    if (
+      typeof obj[key] === 'object' &&
+      obj[key] !== null &&
+      !Array.isArray(obj[key])
+    ) {
+      flattenObject(obj[key] as unknown as FlattenedObject, newKey, result);
+    } else {
+      appendResult[newKey] = obj[key];
+    }
+  });
+  return result;
+}
```

Comment:
The implementation is sound but the typing is iffy, yeah. The input is already assigned to FlattenedObject – assumed to be an object with only string values – which doesn't make sense since the whole point of the function is to flatten nested objects onto one level.

Here's how I would type this. The effective FormikTouched and FormikErrors types already are the same shape as the SimpleObject type given, which means the casts in the next function can also be removed altogether.

```ts
type SimpleObjectNode = SimpleObject | Array<SimpleObjectNode> | string | number | boolean | null;
type SimpleObject = { [prop: string]: SimpleObjectNode };

function isSimpleObject(a: SimpleObjectNode): a is SimpleObject {
  return typeof a === 'object' && a !== null && !Array.isArray(a);
}

function flattenObject(
  obj: SimpleObject,
  parentKey = '',
  result: { [prop: string]: Exclude<SimpleObjectNode, SimpleObject> } = {}
): SimpleObjectNode {
  const appendResult = result;
  Object.keys(obj).forEach((key) => {
    const newKey = `${parentKey}${parentKey ? '.' : ''}${key}`;
    const item = obj[key];
    if (isSimpleObject(item)) {
      flattenObject(item, newKey, result);
    } else {
      appendResult[newKey] = item;
    }
  });
  return result;
}
```

In our use case, where we only care about the object-ness of each leaf and the keys of the output, we could even define SimpleObjectNode as `SimpleObject | unknown` and I believe it would still work.